File: src/main/java/br/com/controle/financeiro/controller/api/TransactionController.java
package br.com.controle.financeiro.controller.api;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import javax.validation.Valid;

import br.com.controle.financeiro.controller.api.linkbuilder.TransactionDTOResourceAssembler;
import br.com.controle.financeiro.model.dto.TransactionDTO;
import br.com.controle.financeiro.model.entity.BankAccount;
import br.com.controle.financeiro.model.entity.Card;
import br.com.controle.financeiro.model.entity.Client;
import br.com.controle.financeiro.model.entity.Transaction;
import br.com.controle.financeiro.model.entity.UserEntity;
import br.com.controle.financeiro.model.exception.TransactionNotFoundException;
import br.com.controle.financeiro.model.repository.BankAccountRepository;
import br.com.controle.financeiro.model.repository.CardRepository;
import br.com.controle.financeiro.model.repository.ClientRepository;
import br.com.controle.financeiro.model.repository.TransactionRepository;
import br.com.controle.financeiro.service.UserService;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.CollectionModel;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping(value = "/api/transaction", produces = MediaType.APPLICATION_JSON_VALUE)
public class TransactionController {

    private static final Logger LOG = LoggerFactory.getLogger(TransactionController.class);

    private final BankAccountRepository accountRepository;
    private final CardRepository cardRepository;
    private final ClientRepository clientRepository;
    private final TransactionRepository transactionRepository;
    private final UserService userService;

    private final TransactionDTOResourceAssembler transactionDTOResourceAssembler;

    public TransactionController(final BankAccountRepository accountRepository, final CardRepository cardRepository,
                                 final ClientRepository clientRepository,
                                 final TransactionRepository transactionRepository, final UserService userService,
                                 final TransactionDTOResourceAssembler transactionDTOResourceAssembler) {
        this.accountRepository = accountRepository;
        this.cardRepository = cardRepository;
        this.clientRepository = clientRepository;
        this.transactionRepository = transactionRepository;
        this.userService = userService;
        this.transactionDTOResourceAssembler = transactionDTOResourceAssembler;
    }

    @GetMapping
    public CollectionModel<EntityModel<TransactionDTO>> allTransactions() {
        LOG.debug("finding allTransactions");
        final UserEntity owner = userService.getAuthenticatedUser();
        final List<EntityModel<TransactionDTO>> transactions =
                transactionRepository.findAllByOwner(owner).stream().map(TransactionDTO::fromTransaction)
                                     .map(transactionDTOResourceAssembler::toModel).collect(Collectors.toList());

        return new CollectionModel<>(transactions,
                               linkTo(methodOn(TransactionController.class).allTransactions()).withSelfRel());
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping(consumes = { MediaType.APPLICATION_JSON_VALUE })
    public EntityModel<TransactionDTO> newTransaction(@RequestBody @Valid final TransactionDTO transaction) {
        LOG.debug("creating newTransaction");
        //TODO deal when an transaction doesn't belong to a card and a bank account simultaneously
        //TODO extract to service
        final UserEntity owner = userService.getAuthenticatedUser();
        Client client = clientRepository.findByIdAndOwner(transaction.getClient(), owner).orElseThrow();
        BankAccount account = accountRepository.findByIdAndOwner(transaction.getAccount(), owner).orElseThrow();
        Card card = cardRepository.findByIdAndOwner(transaction.getCard(), owner).orElseThrow();

        TransactionDTO savedTransaction = TransactionDTO
                .fromTransaction(transactionRepository.save(transaction.toTransaction(client, account, card, owner)));

        return transactionDTOResourceAssembler.toModel(savedTransaction);
    }

    @GetMapping(path = "/{id}")
    public EntityModel<TransactionDTO> oneTransaction(@PathVariable(value = "id") final UUID id) {
        LOG.debug("searching oneTransaction ${}", id);
        final UserEntity owner = userService.getAuthenticatedUser();
        final Transaction transaction =
                transactionRepository.findByIdAndOwner(id, owner).orElseThrow(() -> new TransactionNotFoundException(id));

        return transactionDTOResourceAssembler.toModel(TransactionDTO.fromTransaction(transaction));
    }

    @PutMapping(path = "/{id}")
    public EntityModel<TransactionDTO> replaceTransaction(@RequestBody final TransactionDTO newTransaction,
                                                       @PathVariable final UUID id) {
        LOG.info("replaceTransaction");
        //TODO verify DTO integrity
        final UserEntity owner = userService.getAuthenticatedUser();
        Transaction savedTransaction = transactionRepository.findByIdAndOwner(id, owner).map(transaction -> {
            transaction.setName(newTransaction.getName());
            return transactionRepository.save(transaction);
        }).orElseGet(() -> {
            //TODO extract to service
            Client client = clientRepository.findByIdAndOwner(newTransaction.getClient(), owner).orElseThrow();
            BankAccount account = accountRepository.findByIdAndOwner(newTransaction.getAccount(), owner).orElseThrow();
            Card card = cardRepository.findByIdAndOwner(newTransaction.getCard(), owner).orElseThrow();
            newTransaction.setId(id);

            return transactionRepository.save(newTransaction.toTransaction(client, account, card, owner));
        });

        return transactionDTOResourceAssembler.toModel(TransactionDTO.fromTransaction(savedTransaction));
    }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping(path = "/{id}")
    public void deleteTransaction(@PathVariable final UUID id) {
        LOG.debug("trying to deleteTransaction ${}", id);
        //TODO verify authenticated permission
        transactionRepository.deleteById(id);
    }

}

File: src/main/kotlin/br/com/controle/financeiro/controller/api/TransactionController.kt
package br.com.controle.financeiro.controller.api

import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.http.MediaType
import org.springframework.http.HttpStatus
import org.springframework.hateoas.CollectionModel
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.slf4j.LoggerFactory
import br.com.controle.financeiro.model.repository.BankAccountRepository
import br.com.controle.financeiro.model.repository.CardRepository
import br.com.controle.financeiro.model.repository.ClientRepository
import br.com.controle.financeiro.model.repository.TransactionRepository
import br.com.controle.financeiro.model.repository.UserRepository
import br.com.controle.financeiro.model.dto.TransactionDTO
import br.com.controle.financeiro.model.entity.Transaction
import br.com.controle.financeiro.model.exception.TransactionNotFoundException
import br.com.controle.financeiro.controller.api.linkbuilder.TransactionDTOResourceAssembler
import jakarta.validation.Valid
import java.util.UUID

@RequestMapping(value = ["/api/transaction"], produces = [MediaType.APPLICATION_JSON_VALUE])
class TransactionController(
    private val accountRepository: BankAccountRepository, private val cardRepository: CardRepository,
    private val clientRepository: ClientRepository,
    private val transactionRepository: TransactionRepository, private val userRepository: UserRepository,
    private val transactionDTOResourceAssembler: TransactionDTOResourceAssembler
) {
    @GetMapping
    fun allTransactions(): CollectionModel<EntityModel<TransactionDTO>> {
        LOG.debug("finding allTransactions")
        val owner = userRepository.findAll().iterator().next()
        val transactions = transactionRepository.findAllByOwner(owner).stream().map { t: Transaction ->
            TransactionDTO.fromTransaction(
                t
            )
        }
            .map { entity: TransactionDTO -> transactionDTOResourceAssembler.toModel(entity) }
        .toList()
        return CollectionModel.of(
            transactions,
            WebMvcLinkBuilder.linkTo(WebMvcLinkBuilder.methodOn(TransactionController::class.java).allTransactions())
                .withSelfRel()
        )
    }

    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun newTransaction(@RequestBody transaction: @Valid TransactionDTO): EntityModel<TransactionDTO> {
        LOG.debug("creating newTransaction")
        //TODO deal when an transaction doesn't belong to a card and a bank account simultaneously
        //TODO extract to service
        val owner = userRepository.findAll().iterator().next()
        val client = clientRepository.findByIdAndOwner(transaction.client, owner).orElseThrow()
        val account = accountRepository.findByIdAndOwner(transaction.account, owner).orElseThrow()
        val card = cardRepository.findByIdAndOwner(transaction.card, owner).orElseThrow()
        // val savedTransaction: TransactionDTO = TransactionDTO.fromTransaction(
            // transactionRepository.save(
               // transaction.toTransaction(client, account, card, owner)
            // )
        // )
        return transactionDTOResourceAssembler.toModel(transaction)
    }

    @GetMapping(path = ["/{id}"])
    fun oneTransaction(@PathVariable(value = "id") id: UUID): EntityModel<TransactionDTO> {
        LOG.debug("searching oneTransaction \${}", id)
        val owner = userRepository.findAll().iterator().next()
        val transaction =
            transactionRepository.findByIdAndOwner(id, owner).orElseThrow { TransactionNotFoundException(id) }
        return transactionDTOResourceAssembler.toModel(TransactionDTO.fromTransaction(transaction))
    }

    // @PutMapping(path = ["/{id}"])
    // fun replaceTransaction(
    //     @RequestBody newTransaction: TransactionDTO,
    //     @PathVariable id: UUID
    // ): EntityModel<TransactionDTO> {
    //     LOG.info("replaceTransaction")
    //     //TODO verify DTO integrity
    //     val owner = userRepository.findAll().iterator().next()
    //     val savedTransaction = transactionRepository.findByIdAndOwner(id, owner).map { transaction: Transaction ->
    //         // transaction.name = newTransaction.name
    //         transactionRepository.save(transaction)
    //     }.orElseGet {
    //         //TODO extract to service
    //         val client = clientRepository.findByIdAndOwner(newTransaction.client, owner).orElseThrow()
    //         val account = accountRepository.findByIdAndOwner(newTransaction.account, owner).orElseThrow()
    //         val card = cardRepository.findByIdAndOwner(newTransaction.card, owner).orElseThrow()
    //         newTransaction.id = id
    //         transactionRepository.save(newTransaction.toTransaction(client, account, card, owner))
    //     }
    //     return transactionDTOResourceAssembler.toModel(TransactionDTO.fromTransaction(savedTransaction))
    // }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping(path = ["/{id}"])
    fun deleteTransaction(@PathVariable id: UUID) {
        LOG.debug("trying to deleteTransaction \${}", id)
        //TODO verify authenticated permission
        // transactionRepository.deleteById(id)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(TransactionController::class.java)
    }
}
